src/main/java/io/strimzi/kafka/quotas/CachingVolumeObserver.java

```diff
@@ -0,0 +1,154 @@
+    private final VolumeObserver observer;
+    private final Clock clock;
+    private final Duration entriesValidFor;
+    private final ConcurrentMap<CacheKey, VolumeUsage> cachedObservations;
+    private final LinkedHashMap<String, String> defaultTags;
+    private final Map<CacheKey, Counter> evictionsPerRemoteBroker = new ConcurrentHashMap<>();
+
+    private final Logger log = getLogger(CachingVolumeObserver.class);
+
+    /**
+     * @param observer The downstream observer to be notified after processing
+     * @param clock the clock to use for managing cache expiry
+     * @param entriesValidFor how long cache valid entries for
+     * @param defaultTags tags to be added to all metrics
+     */
+    public CachingVolumeObserver(VolumeObserver observer, Clock clock, Duration entriesValidFor, LinkedHashMap<String, String> defaultTags) {
+        this.observer = observer;
+        this.clock = clock;
+        this.entriesValidFor = entriesValidFor;
+        this.defaultTags = new LinkedHashMap<>(defaultTags);
+        cachedObservations = new ConcurrentHashMap<>();
+        Metrics.newGauge(metricName(CachingVolumeObserver.class, "CachedEntries", defaultTags), new Gauge<>() {
+            @Override
+            public Integer value() {
+                return cachedObservations.size();
+            }
+        });
+    }
+
+    @Override
+    public void observeVolumeUsage(VolumeUsageResult result) {
+        VolumeUsageResult outgoing;
+        if (result.getStatus() == SUCCESS) {
+            if (log.isDebugEnabled()) {
+                log.debug("Caching successful observation. Propagating observation along with cached values where appropriate to the next observer.");
+            }
+            outgoing = cacheAndAugment(result);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Not caching failed observation. Propagating failed observation to the next observer.");
+            }
+            maybeExpireCachedObservations();
+            outgoing = result;
+        }
+        observer.observeVolumeUsage(outgoing);
+    }
+
+    private void maybeExpireCachedObservations() {
+        Set<CacheKey> toRemove = cachedObservations.entrySet().stream()
+                .filter(e -> isExpired(e.getValue())).map(Map.Entry::getKey).collect(Collectors.toSet());
+        toRemove.forEach(this::evict);
+    }
+
+    private void evict(CacheKey key) {
+        final Counter counter = getEvictionCounter(key);
+        counter.inc();
+        if (log.isDebugEnabled()) {
+            log.debug("evicting entry logDir: {} on broker: {}", key.logDir, key.brokerId);
+        }
+        cachedObservations.remove(key);
+    }
+
+    private Counter getEvictionCounter(CacheKey key) {
+        return evictionsPerRemoteBroker.computeIfAbsent(key, key1 -> buildCounterForCacheKey("LogDirEvictions", key1));
+    }
+
+    private Counter buildCounterForCacheKey(String name, CacheKey cacheKey) {
```

Comment:
This appears to have only the one call site. We could drop the `name` parameter, since it's always the same, and call the method `buildEvictionCounter()`

Reply:
Good point. A hang over from a plan to have a counter for adding to the cache however we realised that was point less as every update would "add" to the cache. So we switched to a gauge of the current cache size. 

A separate counter for evictions still felt useful as that is indicative of a problem.

Reply:
resolved with 1b1bc93e3563213cd4714e2b3ecd8b536ac6aae5